src/typeInference/annotateAST.ts

```diff
@@ -0,0 +1,158 @@
+import { SequenceMultiplicity, SequenceType, ValueType } from '../expressions/dataTypes/Value';
+import StaticContext from '../expressions/StaticContext';
+import astHelper, { IAST } from '../parsing/astHelper';
+import { annotateBinOp } from './annotateBinaryOperator';
+import { annotateCastableOperator, annotateCastOperator } from './annotateCastOperators';
+import {
+	annotateGeneralCompare,
+	annotateNodeCompare,
+	annotateValueCompare,
+} from './annotateCompareOperator';
+import { annotateFunctionCall } from './annotateFunctionCall';
+import { annotateLogicalOperator } from './annotateLogicalOperator';
+import { annotateUnaryMinus, annotateUnaryPlus } from './annotateUnaryOperator';
+
+/**
+ * Recursively traverse the AST in the depth first, pre-order to infer type and annotate AST;
+ * Annotates as much type information as possible to the AST nodes.
+ * Inserts attribute `type` to the corresponding node if type is inferred.
+ *
+ * @param ast The AST to annotate
+ * @param staticContext The static context used for function lookups
+ */
+export default function annotateAst(ast: IAST, staticContext?: StaticContext) {
+	annotate(ast, staticContext);
+}
+
+/**
+ * Recursively traverse the AST in the depth first, pre-order to infer type and annotate AST;
+ * Annotates as much type information as possible to the AST nodes.
+ * Inserts attribute `type` to the corresponding node if type is inferred.
+ *
+ * @param ast The AST to annotate.
+ * @param staticContext The static context to use for function lookups.
+ * @throws errors when attempts to annotate fail.
+ * @returns The type of the AST node or `undefined` when the type cannot be annotated.
```

Comment:
Or it can return `item()*`, since that is the  most generic type. That way you can remove all of the checks for `undefined` I assume you guys have all over the place